File: solver/app/schemas.py
"""Pydantic 입출력 스키마. Go side `internal/solver/schemas.go`와 JSON 호환.

Design Ref: §4.3 솔버 API
"""

import datetime as _dt
from datetime import date
from typing import List, Literal, Optional

from pydantic import BaseModel, Field, field_validator

ShiftCode = Literal["D", "E", "N", "O", "DE"]
WishType = Literal["off", "d", "e", "n", "unavailable"]
FixedPattern = Literal["D_ONLY", "E_ONLY", "N_ONLY", "WEEKDAY_D", "WEEKDAY_E"]


# ============================================================
# Input
# ============================================================


class NurseIn(BaseModel):
    id: int
    level: str                                # experience_levels.code (G-04)
    fixed_pattern: Optional[FixedPattern] = None  # H-11
    is_night_keeper: bool = False             # K-01


class WishIn(BaseModel):
    nurse_id: int
    date: date
    type: WishType


class PrevCellIn(BaseModel):
    """이전 달 마지막 lookback_days 일치 cells (H-10). 솔버 상수로 취급."""

    nurse_id: int
    date: date
    shift: ShiftCode


class ExperienceLevelIn(BaseModel):
    code: str
    min_d: int = 0
    min_e: int = 0
    min_n: int = 0
    weight_coverage: int = 1
    weight_d_assignment: int = 0
    weight_e_assignment: int = 0
    weight_n_assignment: int = 0


class WardSettingsIn(BaseModel):
    min_d: int = 3
    min_e: int = 2
    min_n: int = 2
    max_consecutive_n: int = 3
    min_rest_after_n: int = 1
    max_consecutive_workdays: int = 5
    balance_off_tolerance: int = 1
    previous_month_lookback_days: int = 7  # H11 fix: H-10
    weight_balance_off: int = 10
    weight_respect_wishes: int = 8
    weight_weekend_balance: int = 5
    weight_same_shift_streak: int = 3
    weight_short_rest_pattern: int = 4


class GenerateInput(BaseModel):
    """POST /generate 요청. Design §4.3."""

    year_month: str = Field(pattern=r"^\d{4}-\d{2}$")
    nurses: List[NurseIn]
    wishes: List[WishIn] = []
    previous_month_last_week_cells: List[PrevCellIn] = []  # H-10
    experience_levels: List[ExperienceLevelIn]
    ward_settings: WardSettingsIn = WardSettingsIn()
    max_time_seconds: int = 55

    @field_validator("nurses")
    @classmethod
    def _nurses_unique(cls, v):
        ids = [n.id for n in v]
        if len(set(ids)) != len(ids):
            raise ValueError("nurses에 중복 id가 있습니다")
        return v


class ValidateInput(BaseModel):
    """POST /validate 요청. 수동 수정 검증용."""

    year_month: str = Field(pattern=r"^\d{4}-\d{2}$")
    nurses: List[NurseIn]
    wishes: List[WishIn] = []
    previous_month_last_week_cells: List[PrevCellIn] = []
    experience_levels: List[ExperienceLevelIn]
    ward_settings: WardSettingsIn = WardSettingsIn()
    all_cells: List["CellOut"]


# ============================================================
# Output
# ============================================================


class CellOut(BaseModel):
    nurse_id: int
    date: date
    shift: ShiftCode


class GenerateOutput(BaseModel):
    status: Literal["ok", "infeasible", "timeout", "error"]
    solver_status: str
    objective_value: Optional[int] = None
    cells: List[CellOut] = []
    applied_rules: List[str] = []
    elapsed_ms: int
    violated_rule_ids: List[str] = []
    suggestion: Optional[str] = None


class Violation(BaseModel):
    rule_id: str
    severity: Literal["hard", "soft"]
    message: str
    cell_ids: List[int] = []
    nurse_id: Optional[int] = None
    # field 이름 'date'와 datetime.date의 self-shadow 회피를 위해 module 경로 사용
    date: Optional[_dt.date] = None


class ValidateOutput(BaseModel):
    violations: List[Violation]
    hard_count: int
    soft_count: int


ValidateInput.model_rebuild()

File: api/internal/solver/schemas.go
// Package solver — Python solver와 1:1 JSON 스키마.
// Design Ref: §4.3, §9.1 (solver 슬라이스)
//
// solver/app/schemas.py의 Pydantic 모델과 필드명 일치해야 함.
package solver

import "time"

type NurseIn struct {
	ID            int     `json:"id"`
	Level         string  `json:"level"`
	FixedPattern  *string `json:"fixed_pattern,omitempty"` // H-11
	IsNightKeeper bool    `json:"is_night_keeper"`         // K-01
}

type WishIn struct {
	NurseID int       `json:"nurse_id"`
	Date    time.Time `json:"date"` // YYYY-MM-DD 마샬
	Type    string    `json:"type"` // off|d|e|n|unavailable
}

type PrevCellIn struct {
	NurseID int       `json:"nurse_id"`
	Date    time.Time `json:"date"`
	Shift   string    `json:"shift"` // D|E|N|O|DE
}

type ExperienceLevelIn struct {
	Code              string `json:"code"`
	MinD              int    `json:"min_d"`
	MinE              int    `json:"min_e"`
	MinN              int    `json:"min_n"`
	WeightCoverage    int    `json:"weight_coverage"`
	WeightDAssignment int    `json:"weight_d_assignment"`
	WeightEAssignment int    `json:"weight_e_assignment"`
	WeightNAssignment int    `json:"weight_n_assignment"`
}

type WardSettingsIn struct {
	MinD                     int `json:"min_d"`
	MinE                     int `json:"min_e"`
	MinN                     int `json:"min_n"`
	MaxConsecutiveN          int `json:"max_consecutive_n"`
	MinRestAfterN            int `json:"min_rest_after_n"`
	MaxConsecutiveWorkdays   int `json:"max_consecutive_workdays"`
	BalanceOffTolerance      int `json:"balance_off_tolerance"`
	WeightBalanceOff         int `json:"weight_balance_off"`
	WeightRespectWishes      int `json:"weight_respect_wishes"`
	WeightWeekendBalance     int `json:"weight_weekend_balance"`
	WeightSameShiftStreak    int `json:"weight_same_shift_streak"`
	WeightShortRestPattern   int `json:"weight_short_rest_pattern"`
}

// ----- Input -----

type GenerateInput struct {
	YearMonth                  string              `json:"year_month"`
	Nurses                     []NurseIn           `json:"nurses"`
	Wishes                     []WishIn            `json:"wishes"`
	PreviousMonthLastWeekCells []PrevCellIn        `json:"previous_month_last_week_cells"` // H-10
	ExperienceLevels           []ExperienceLevelIn `json:"experience_levels"`
	WardSettings               WardSettingsIn      `json:"ward_settings"`
	MaxTimeSeconds             int                 `json:"max_time_seconds"`
}

type CellOut struct {
	NurseID int       `json:"nurse_id"`
	Date    time.Time `json:"date"`
	Shift   string    `json:"shift"`
}

type ValidateInput struct {
	YearMonth                  string              `json:"year_month"`
	Nurses                     []NurseIn           `json:"nurses"`
	Wishes                     []WishIn            `json:"wishes"`
	PreviousMonthLastWeekCells []PrevCellIn        `json:"previous_month_last_week_cells"`
	ExperienceLevels           []ExperienceLevelIn `json:"experience_levels"`
	WardSettings               WardSettingsIn      `json:"ward_settings"`
	AllCells                   []CellOut           `json:"all_cells"`
}

// ----- Output -----

type GenerateOutput struct {
	Status           string    `json:"status"` // ok|infeasible|timeout|error
	SolverStatus     string    `json:"solver_status"`
	ObjectiveValue   *int      `json:"objective_value,omitempty"`
	Cells            []CellOut `json:"cells"`
	AppliedRules     []string  `json:"applied_rules"`
	ElapsedMs        int       `json:"elapsed_ms"`
	ViolatedRuleIDs  []string  `json:"violated_rule_ids,omitempty"`
	Suggestion       *string   `json:"suggestion,omitempty"`
}

type Violation struct {
	RuleID   string     `json:"rule_id"`
	Severity string     `json:"severity"` // hard|soft
	Message  string     `json:"message"`
	CellIDs  []int      `json:"cell_ids,omitempty"`
	NurseID  *int       `json:"nurse_id,omitempty"`
	Date     *time.Time `json:"date,omitempty"`
}

type ValidateOutput struct {
	Violations []Violation `json:"violations"`
	HardCount  int         `json:"hard_count"`
	SoftCount  int         `json:"soft_count"`
}
